### PR TITLE
Add support for enhanced error messages.

### LINKED ISSUE #109: Support multiple messages for single character based rules
It would be useful to allow some rules to leverage more than a single message based on its failure.
In the case of `AllowedCharacterRule` and `IllegalCharacterRule` it's natural to want a different message based on whitespace.
It can also be hard to construct a single message for all character types.

### PR DESCRIPTION
**Pull request by dfish3r**
Update AllowedCharacterRule and IllegalCharacterRule to support per-character error messaging.

Fixes #109 .

### CODE REVIEW

**Comment by dfish3r** [comment]
Note that changes to `PropertiesMessageResolver` are potentially breaking to anyone who has implemented custom error code keys that include dot (.)

**Comment by serac** [comment]
I think folks tend to paint in the existing lines when extending. Since the default error code keys don't contain dots, I would think consumers would naturally avoid them. That's to say I think the impact will be negligible to zero.

File: src/main/java/org/passay/PropertiesMessageResolver.java [hunk 1]
```diff
@@ -45,7 +45,15 @@ public PropertiesMessageResolver(final Properties properties)
   @Override
   protected String getMessage(final String key)
   {
-    return messageProperties.getProperty(key);
+    String msg = messageProperties.getProperty(key);
+    if (msg == null && key.contains(".")) {
+      String newKey = key;
+      while (msg == null && newKey.contains(".")) {
```
Comment (serac) [inline src/main/java/org/passay/PropertiesMessageResolver.java]: Implementing this as a loop suggests that you intend to support multiple dots, which is not what we discussed. What's your thinking here?

File: src/main/java/org/passay/AllowedCharacterRule.java [hunk 2]
```diff
@@ -115,7 +148,11 @@ public RuleResult validate(final PasswordData passwordData)
     for (char c : text.toCharArray()) {
       if (Arrays.binarySearch(allowedCharacters, c) < 0 && !matches.contains(c)) {
         if (MatchBehavior.Contains.equals(matchBehavior) || matchBehavior.match(text, c)) {
-          result.addError(ERROR_CODE, createRuleResultDetailParameters(c));
+          if (enhancedErrorMessages) {
```
[the remark this answers is not in the document]
Comment (dfish3r) [reply src/main/java/org/passay/AllowedCharacterRule.java]: No. It's insurance against people using resolvers other than `PropertiesMessageResolver`. In fact, we supply one called `SpringMessageResolver`. If this rule changes its default behavior, other message resolvers will break.

File: src/main/java/org/passay/PropertiesMessageResolver.java (hunk 1 above)
Comment (dfish3r) [reply src/main/java/org/passay/PropertiesMessageResolver.java]: Seemed like a general improvement to support hierarchical messages. What's the downside?

File: src/main/java/org/passay/AllowedCharacterRule.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (serac) [reply src/main/java/org/passay/AllowedCharacterRule.java]: An alternative just came to me: emit multiple error codes per message. Then you could implement a "most-specific" resolution strategy, but custom resolvers would still have the "old" code around for backward-compatible message resolution.

[the remark this answers is not in the document]
Comment (dfish3r) [reply src/main/java/org/passay/AllowedCharacterRule.java]: The `PropertiesMessageResolver` emits a default message if no message is resolved. Custom resolvers may do the same thing. I think changing this behavior is potentially breaking. We can do it, but a minor version bump would be in order IMO.

File: src/main/java/org/passay/PropertiesMessageResolver.java (hunk 1 above)
Comment (serac) [reply src/main/java/org/passay/PropertiesMessageResolver.java]: None, just wanted to poke at what you were thinking a bit.

File: src/main/java/org/passay/AllowedCharacterRule.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (serac) [reply src/main/java/org/passay/AllowedCharacterRule.java]: I'm going to spend an hour taking a crack at it.

[the remark this answers is not in the document]
Comment (serac) [reply src/main/java/org/passay/AllowedCharacterRule.java]: Sent you a PR with an implementation that does what I want. It appears to be wholly backward compatible. https://github.com/vt-middleware/passay/pull/111

[the remark this answers is not in the document]
Comment (serac) [reply src/main/java/org/passay/AllowedCharacterRule.java]: PR accepted

**Pull request merged**